// father/lib/doc/doczrc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _doczPluginUmiCss = require("docz-plugin-umi-css");

var _path = require("path");

var _fs = require("fs");

var _lodash = require("lodash");

var _getUserConfig = _interopRequireWildcard(require("father-build/lib/getUserConfig"));

var _registerBabel = _interopRequireDefault(require("father-build/lib/registerBabel"));

var _doczPluginReactExternals = _interopRequireDefault(require("./docz-plugin-react-externals"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const cssModuleRegex = /\.module\.css$/;
const lessModuleRegex = /\.module\.less$/;
const sassModuleRegex = /\.module\.[(sass)|(scss)]$/;
const cwd = process.cwd();
const localUserConfig = JSON.parse((0, _fs.readFileSync)((0, _path.join)(cwd, '.docz', '.fatherrc.json'), 'utf-8')); // register babel for config files

(0, _registerBabel.default)({
  cwd,
  only: _getUserConfig.CONFIG_FILES
});

const userConfig = _objectSpread(_objectSpread({}, localUserConfig), (0, _getUserConfig.default)({
  cwd
}));

if (!userConfig.doc) {
  userConfig.doc = (0, _lodash.merge)(userConfig.doc || {});
}

const isTypescript = (0, _fs.existsSync)((0, _path.join)(cwd, 'tsconfig.json'));

var _default = _objectSpread(_objectSpread({
  typescript: isTypescript,
  repository: false,
  theme: require.resolve('docz-theme-umi').replace(/\\/g, '/')
}, userConfig.doc), {}, {
  modifyBabelRc(babelrc, args) {
    if (typeof userConfig.doc.modifyBabelRc === 'function') {
      babelrc = userConfig.doc.modifyBabelRc(babelrc, args);
    } // 需放 class-properties 前面


    babelrc.plugins.unshift([require.resolve('@babel/plugin-proposal-decorators'), {
      legacy: true
    }]); // Support extraBabelPresets and extraBabelPlugins

    babelrc.presets = [...babelrc.presets, ...(userConfig.extraBabelPresets || [])];
    babelrc.plugins = [...babelrc.plugins, ...(userConfig.extraBabelPlugins || [])];
    return babelrc;
  },

  modifyBundlerConfig(config, dev, args) {
    if (userConfig.doc.modifyBundlerConfig) {
      config = userConfig.doc.modifyBundlerConfig(config, dev, args);
    }

    if (!dev) {
      // do not generate doc sourcemap
      config.devtool = false; // support disable minimize via process.env.COMPRESS

      if (process.env.COMPRESS === 'none') {
        config.optimization.minimize = false;
      }
    } // 确保只有一个版本的 docz，否则 theme 会出错，因为 ComponentProvider 的 context 不是同一个


    config.resolve.alias = config.resolve.alias || {};
    config.resolve.alias.docz = (0, _path.dirname)(require.resolve('docz/package.json')); // 透传 BIGFISH_VERSION 环境变量

    config.plugins.push(new (require('webpack').DefinePlugin)({
      'process.env.BIGFISH_VERSION': JSON.stringify(process.env.BIGFISH_VERSION)
    })); // fallback resolve 路径

    config.resolve.modules.push((0, _path.join)(__dirname, '../../node_modules'));
    config.resolveLoader.modules.push((0, _path.join)(__dirname, '../../node_modules'));
    return config;
  },

  plugins: [...(userConfig.doc.plugins || []), (0, _doczPluginReactExternals.default)(), ...(userConfig.cssModules ? [// .css
  (0, _doczPluginUmiCss.css)({
    preprocessor: 'postcss',
    ruleOpts: {
      exclude: /node_modules\/.*\.css$/
    },
    cssmodules: true
  }), (0, _doczPluginUmiCss.css)({
    preprocessor: 'postcss',
    ruleOpts: {
      test: /node_modules\/.*\.css$/
    },
    cssmodules: false
  }), // .less
  (0, _doczPluginUmiCss.css)({
    preprocessor: 'less',
    ruleOpts: {
      exclude: /node_modules\/.*\.less$/
    },
    cssmodules: true,
    loaderOpts: _objectSpread({
      javascriptEnabled: true
    }, userConfig.lessInRollupMode)
  }), (0, _doczPluginUmiCss.css)({
    preprocessor: 'less',
    ruleOpts: {
      test: /node_modules\/.*\.less$/
    },
    cssmodules: false,
    loaderOpts: _objectSpread({
      javascriptEnabled: true
    }, userConfig.lessInRollupMode)
  }), // .sass
  (0, _doczPluginUmiCss.css)({
    preprocessor: 'sass',
    ruleOpts: {
      exclude: /node_modules\/.*\.sass$/
    },
    cssmodules: true,
    loaderOpts: _objectSpread({}, userConfig.sassInRollupMode)
  }), (0, _doczPluginUmiCss.css)({
    preprocessor: 'sass',
    ruleOpts: {
      test: /node_modules\/.*\.sass$/
    },
    cssmodules: false,
    loaderOpts: _objectSpread({}, userConfig.sassInRollupMode)
  })] : [// .css
  (0, _doczPluginUmiCss.css)({
    preprocessor: 'postcss',
    ruleOpts: {
      exclude: cssModuleRegex
    },
    cssmodules: false
  }), (0, _doczPluginUmiCss.css)({
    preprocessor: 'postcss',
    ruleOpts: {
      test: cssModuleRegex
    },
    cssmodules: true
  }), // .less
  (0, _doczPluginUmiCss.css)({
    preprocessor: 'less',
    ruleOpts: {
      exclude: lessModuleRegex
    },
    cssmodules: false,
    loaderOpts: _objectSpread({
      javascriptEnabled: true
    }, userConfig.lessInRollupMode)
  }), (0, _doczPluginUmiCss.css)({
    preprocessor: 'less',
    ruleOpts: {
      test: lessModuleRegex
    },
    cssmodules: true,
    loaderOpts: _objectSpread({
      javascriptEnabled: true
    }, userConfig.lessInRollupMode)
  }), // sass
  (0, _doczPluginUmiCss.css)({
    preprocessor: 'sass',
    ruleOpts: {
      exclude: sassModuleRegex
    },
    cssmodules: false,
    loaderOpts: _objectSpread({}, userConfig.sassInRollupMode)
  }), (0, _doczPluginUmiCss.css)({
    preprocessor: 'sass',
    ruleOpts: {
      test: sassModuleRegex
    },
    cssmodules: true,
    loaderOpts: _objectSpread({}, userConfig.sassInRollupMode)
  })])]
});

exports.default = _default;
